### PR TITLE
added solution

### PR DESCRIPTION
**Pull request by anatolii-zarytskyi**

### CODE REVIEW

File: src/main/java/core/basesyntax/Car.java [hunk 1]
```diff
@@ -1,13 +1,86 @@
...
+    private final String color;
+    private final List<Wheel> wheels;
+    private final Engine engine;
+
+    public Car(int year, String color, List<Wheel> wheels, Engine engine) {
+        this.year = year;
+        this.color = color;
+        this.wheels = cloneWheels(wheels);
+        this.engine = engine == null ? null : engine.clone();
+    }
```
Comment (Aalderaan) [inline src/main/java/core/basesyntax/Car.java:18]: This  repeats two times, maybe it is good to create method, that check for null and clone engine?

File: src/main/java/core/basesyntax/Car.java [hunk 2]
```diff
@@ -1,13 +1,86 @@
...
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(year, color, wheels, engine);
+    }
+
+    private List<Wheel> cloneWheels(List<Wheel> wheels) {
+        List<Wheel> clonedWheels = new ArrayList<>();
+        for (Wheel wheel : wheels) {
+            clonedWheels.add(new Wheel(wheel.getRadius()));
```
Comment (Aalderaan) [inline src/main/java/core/basesyntax/Car.java:82]: Maybe It is possible to make it shorter - (similar is on line 46) by adding directly wheel.clon()?

**Review by Aalderaan — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/Car.java (hunk 1 above)
Comment (Anatolii-Zarytskyi) [reply src/main/java/core/basesyntax/Car.java]: done

File: src/main/java/core/basesyntax/Car.java (hunk 2 above)
Comment (Anatolii-Zarytskyi) [reply src/main/java/core/basesyntax/Car.java]: Agree. Done.

**Review by Aalderaan — CHANGES_REQUESTED** [review]

**Review by Aalderaan — APPROVED** [review]

File: src/main/java/core/basesyntax/Car.java [hunk 3]
```diff
@@ -1,13 +1,90 @@
...
+        car.wheels.add(newWheel.clone());
+        return car;
+    }
+
+    private Engine checkAndCloneEngine(Engine engine) {
+        return engine == null ? null : engine.clone();
+    }
+
+    @Override
+    public Car clone() {
```
Comment (RomanDubovskyi) [inline src/main/java/core/basesyntax/Car.java:55]: we don't need this method, besides it doesn't create deep copy now  - let's remove it

File: src/main/java/core/basesyntax/Car.java [hunk 4]
```diff
@@ -1,13 +1,90 @@
...
+        return new Car(year, color, wheels, engine);
+    }
+
+    public Car addWheel(Wheel newWheel) {
+        Car car = new Car(year, color, wheels, engine);
+        car.wheels.add(newWheel.clone());
+        return car;
+    }
+
+    private Engine checkAndCloneEngine(Engine engine) {
```
Comment (RomanDubovskyi) [inline src/main/java/core/basesyntax/Car.java:50]: move this method near the public method where you use it or to the end of the file near to your second private method

**Review by RomanDubovskyi — CHANGES_REQUESTED** [review]

**Review by RomanDubovskyi — APPROVED** [review]

**Pull request closed without merging**